src/compiler/emitter.ts

```diff
@@ -5657,6 +5869,91 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
+                switch (compilerOptions.jsx) {
+                    case JsxEmit.React:
+                        jsxEmitReact(node);
+                        break;
+                    // Fall back to preserve if None was specified (we'll error earlier)
+                    default:
+                        jsxEmitPreserve(node);
+                        break;
+                }
+            }
+
+            function trimReactWhitespace(node: JsxText): string {
+                // Could be empty string, do not use !node.formattedReactText
+                if (node.formattedReactText !== undefined) {
+                    return node.formattedReactText;
+                }
+
+                let lines: string[] = [];
+                let text = getTextOfNode(node);
+                let firstNonWhitespace = 0;
+                let lastNonWhitespace = -1;
+
+                for (var i = 0; i < text.length; i++) {
+                    var c = text.charCodeAt(i);
+                    if (c === CharacterCodes.lineFeed || c === CharacterCodes.carriageReturn) {
+                        if (firstNonWhitespace !== -1 && (lastNonWhitespace - firstNonWhitespace + 1 > 0)) {
+                            lines.push(text.substr(firstNonWhitespace, lastNonWhitespace - firstNonWhitespace + 1));
+                        }
+                        firstNonWhitespace = -1;
+                    }
+                    else if (!isWhiteSpace(c)) {
+                        lastNonWhitespace = i;
+                        if (firstNonWhitespace === -1) {
+                            firstNonWhitespace = i;
+                        }
+                    }
+                }
+                if (firstNonWhitespace !== -1) {
+                    lines.push(text.substr(firstNonWhitespace));
+                }
+
+                return node.formattedReactText = lines.join('" + \' \' + "');
+            }
+
+            function shouldEmitJsxText(node: JsxText) {
+                if (compilerOptions.jsx === JsxEmit.React) {
+                    return trimReactWhitespace(node).length > 0;
+                }
+                else {
+                    return true;
+                }
+            }
+
+            function emitJsxText(node: JsxText) {
+                switch (compilerOptions.jsx) {
+                    case JsxEmit.React:
+                        write('"');
+                        write(trimReactWhitespace(node));
+                        write('"');
+                        break;
+
+                    case JsxEmit.Preserve:
+                    default: // Emit JSX-preserve as default when no --jsx flag is specified
+                        write(getTextOfNode(node, true));
+                        break;
+                }
+            }
+
+            function emitJsxExpression(node: JsxExpression) {
+                if (node.expression && node.expression.kind !== SyntaxKind.OmittedExpression) {

```

Comment:
Is an omitted expression syntactically valid and bound to occur in any scenario?